### PR TITLE
fix: Rules with regex and begin-end selectors replace all occurences of matched value

### PR DESCRIPTION
**Pull request by going-confetti**
Closes https://github.com/grafana/k6-studio/issues/848

## Description

> [!NOTE]  
> Empty regex expressions/expressions with empty groups are supported as well as begin-end pairs that match empty strings. This is an intentional change, but do let me know if there's something I missed when making it.

- Change the replacement logic used in rule preview to fix the issue
- Add tests for this edge case
- Refactor the code across `rules/correlation.utils.ts`, `rules/shared.ts`, and `rules/selectors/text.ts`

grafik

## How to Test

> [!IMPORTANT]  
> Make sure to test this thoroughly and let me know if you think something should be covered by tests

1. Verify that https://github.com/grafana/k6-studio/issues/848 is addressed (also check that the generated test works as expected
2. Verify that the following functionality works as expected:
    - Rule preview for correlation and parameterization rules
    - Correlation rules with- and without custom replacers
    - All types of selectors (begin-end, regex, header-name, JSON, text) in all kinds of targets (URL, headers, body)

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added tests for my changes.
- [x] I have commented on my code, particularly in hard-to-understand areas.

### CODE REVIEW

File: src/rules/correlation.utils.ts [hunk 1]
```diff
@@ -35,31 +37,19 @@ export function replaceCorrelatedValues({
...
-    extractedValue,
-    variableName
-  )
-  const cookies: Cookie[] = replaceCookies(
-    request.cookies,
-    extractedValue,
-    variableName
-  )
+  oldValue: string,
+  newValue: string
```
Comment (going-confetti) [inline src/rules/correlation.utils.ts:41]: Should I have kept `variableName` here? I figured it wasn't important in this context, but I'm not sure anymore.

File: src/rules/shared.ts [hunk 2]
```diff
@@ -97,51 +97,39 @@ export const getJsonObjectFromPath = (json: string, path: string) => {
...
-  headers: Header[],
-  value: string,
-  newValue: string
-): Header[] => {
-  return headers.map(([key, headerValue]) => {
-    const replacedValue = headerValue.replaceAll(value, newValue)
-    return [key, replacedValue]
-  })
+  const regex = new RegExp(`${escapeRegExp(begin)}(.*?)${escapeRegExp(end)}`)
+  return content.replace(regex, `${begin}${newValue}${end}`)
```
Comment (going-confetti) [inline src/rules/shared.ts]: This is enough to replace the first occurrence of the extracted value surrounded within the begin-end boundaries.

File: src/rules/shared.ts [hunk 3]
```diff
@@ -97,51 +97,39 @@ export const getJsonObjectFromPath = (json: string, path: string) => {
...
+    // If no capture group, replace the entire match
+    return newValue
   })
 }
 
-export const replaceBeginEndBody = (
+const replaceBeginEndBody = (
   selector: BeginEndSelector,
   request: Request,
   variableName: string
```
Comment (going-confetti) [inline src/rules/shared.ts]: Here, `variableName` does not look like the correct param name (it only does when this is used for correlation). This is out of scope of this PR, but since I've already made so many changes here, should I rename it to `value`?

File: src/rules/shared.ts [hunk 4]
```diff
@@ -284,318 +307,77 @@ const replaceHeaderByName = (
...
-      scheme: 'http',
-      host: 'localhost:3000',
-      content,
-      path: '/api/v1/foo',
-      timestampStart: 0,
-      timestampEnd: 0,
-      contentLength: 0,
-      httpVersion: '1.1',
-    }
+export function replaceAllBody(
```
Comment (going-confetti) [inline src/rules/shared.ts]: `replaceAllBody`, `replaceAllUrl`, and others are basically replacements for the old `replaceContent` `replaceURL`, etc. with a few minor changes (such as their names being more explicit about what the function does (although at this point I'd prefer all names to follow the "replace X in Y pattern", but that's a separate discussion), and the return type being `Request`).

File: src/rules/shared.ts [hunk 5]
```diff
@@ -97,155 +98,189 @@ export const getJsonObjectFromPath = (json: string, path: string) => {
...
+const replaceRegexUrl = (
   selector: RegexSelector,
   request: Request,
   variableName: string
-) => {
-  const valueToReplace = matchRegex(request.url, selector.regex)
-  if (!valueToReplace) return
+): Request | undefined => {
+  const match = matchRegex(request.url, selector.regex)
+  if (match === undefined) return
```
Comment (going-confetti) [inline src/rules/shared.ts]: IMO it's perfectly valid to want to use an empty group in some cases. Off the top of my head, I can think of a few examples, such as:
- Add a new query param
- Modify the URL to include prefix/suffix depending on the environment, etc.
- Parameterize emails for plus-addressing (e.g. `()@`)

Image

Image

The downside is that it now matches the beginning of the string (see the screenshot below). I'd argue it's actually good because it provides immediate feedback (whereas before if you had an empty regex, the rule would not work) - but I realize that I might be biased.

Image

File: src/rules/correlation.utils.ts (hunk 1 above)
Comment (e-fisher) [reply src/rules/correlation.utils.ts:41]: Old/new makes more sense in this context 👍

File: src/rules/shared.ts (hunk 5 above)
Comment (e-fisher) [reply src/rules/shared.ts]: I think empty regex input shouldn't produce any matches - it just gives an impression of a bug to me. Do you know any examples where this would be useful?

Comment (going-confetti) [reply src/rules/shared.ts]: Yeah, the more I think about it, the more it looks like a bug. I fixed this 👍

File: src/rules/selectors/beginEnd.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    case 'headers':
+      return replaceBeginEndHeaders(selector, request, newValue)
+    case 'url':
+      return replaceBeginEndUrl(selector, request, newValue)
+    default:
+      return exhaustive(selector.from)
+  }
+}
+
+export function matchBeginEnd(value: string, begin: string, end: string) {
```
Comment (going-confetti) [inline src/rules/selectors/beginEnd.ts:24]: Petition to rename all these `matchX` functions to `extractX` 👀

File: src/rules/selectors/beginEnd.ts
Comment (going-confetti) [inline src/rules/selectors/beginEnd.ts:1]: Oops, splitting `shared.ts` into multiple files made it harder to review the changes, I apologize.

File: src/views/Generator/RuleEditor/SelectorField.tsx [hunk 7]
```diff
@@ -187,7 +187,17 @@ function SelectorContent({
...
     case 'regex':
       return (
-        <FieldGroup name={`${field}.regex`} errors={errors} label="Regex">
+        <FieldGroup
+          name={`${field}.regex`}
+          errors={errors}
+          label="Regex"
+          hint={
+            <Text size="1">
+              First match only. Use a capture group (e.g.,{' '}
```
Comment (going-confetti) [inline src/views/Generator/RuleEditor/SelectorField.tsx:196]: Image
